pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,369 @@
+	"k8s.io/kubernetes/pkg/volume/fc"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/kubernetes/fake"
+	utiltesting "k8s.io/client-go/util/testing"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetest "k8s.io/kubernetes/pkg/volume/testing"
+)
+
+const (
+	testHostName = "test-hostname"
+	socketPath   = "/var/run/kmsplugin"
+	volumeName   = "volume-name"
+	volumeName2  = "volume-name-2"
+	testNodeName = "test-node-name"
+)
+
+var (
+	emptyVolume           = v1.Volume{}
+	emptyPersistentVolume = v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{},
+	}
+	dirOrCreate = v1.HostPathType(v1.HostPathDirectoryOrCreate)
+	nodeName    = kubetypes.NodeName(testNodeName)
+	hostPath    = &v1.HostPathVolumeSource{
+		Path: socketPath,
+		Type: &dirOrCreate,
+	}
+	objectReference  = v1.ObjectReference{Namespace: "claimrefns", Name: "test-pvc"}
+	objectReference2 = v1.ObjectReference{Namespace: "claimrefns", Name: "test-pv"}
+
+	volumeExample = v1.Volume{
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			HostPath: hostPath,
+		},
+	}
+	volumeExampleWithCSI = v1.Volume{
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+				ClaimName: "test-pvc",
+				ReadOnly:  false,
+			},
+		},
+	}
+	pvExampleWithCSI = v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				CSI: &v1.CSIPersistentVolumeSource{
+					Driver:           "pd.csi.storage.gke.io",
+					VolumeHandle:     "projects/UNSPECIFIED/zones/UNSPECIFIED/disks/test-disk",
+					FSType:           "ext4",
+					ReadOnly:         false,
+					VolumeAttributes: map[string]string{"partition": ""},
+				},
+			},
+			ClaimRef: &objectReference,
+		},
+	}
+)
+
+type vaTest struct {
+	desc                 string
+	seupNodeName         string
+	createNodeName       kubetypes.NodeName
+	podVolume            v1.Volume
+	pod                  *v1.Pod
+	wantVolume           v1.Volume
+	wantPersistentVolume v1.PersistentVolume
+	wantErrorMessage     string
+}
+
+func Test_CreateVolumeSpec(t *testing.T) {
+	for _, test := range []vaTest{
+		{
+			desc:           "inline volume type that does not support csi migration",
+			seupNodeName:   testNodeName,
+			createNodeName: nodeName,
+			podVolume:      volumeExample,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+							},
+						},
+					},
+				},
+			},
+			wantVolume: volumeExample,
+		},
+		{
+			desc:           "inline volume type that supports csi migration",
+			seupNodeName:   testNodeName,
+			createNodeName: nodeName,
+			podVolume: v1.Volume{
+				Name: volumeName,
+				VolumeSource: v1.VolumeSource{
+					CSI: &v1.CSIVolumeSource{
```

Comment:
this should be an intree volume type like `GCEPersistentDisk`, instead of using `CSI` directly

Reply:
How should the return spec of this look like? Is it a CSIVolumeSource?